src/Indexer.php

```diff
@@ -196,6 +230,35 @@ public function index(): Promise
         });
     }
 
+    /**
+     * Return current indexing state
+     *
+     * @return bool
+     */
+    public function isIndexing(): bool
+    {
+        return $this->isIndexing;
+    }
+
+    /**
+     * Cancel all running indexing processes
+     *
+     * @return Promise
+     */
+    public function cancel(): Promise
+    {
+        return coroutine(function () {
+            $this->hasCancellationSignal = true;
+
+            while ($this->isIndexing()) {
+                yield timeout();
+            }
```

Comment:
I'm not that happy with this "polling" solution - usually cancellation is synchronous and best-effort. I.e. the old indexing would stop _eventually_ but you can already start with the next index run, by just making sure to instantiate new Index objects (maybe even a new Indexer) so that the old run can't accidentally still write to the new index. Does that sound feasible?

Reply:
Don't think a new Indexer instance is a good idea, they are used in 2 different places (`initialized` and `didChangeConfiguration`) and I would need a reference to both.

And I think creating new Index objects is too much work from what I see.
Are you talking about `sourceIndex` and `dependencyIndex`?

If yes, then the method would need to know about all objects that require the index objects. Or am I wrong here?